apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts

```diff
@@ -0,0 +1,1029 @@
+
+        await Promise.all(linkPromises);
+
+        return { linkCount };
+      }
+    );
+
+    this._terminal.writeVerboseLine(
+      `${shouldHardlink ? 'Hardlink' : 'Copy from cache'}: ${hardlinkDuration}ms (${hardlinkCount} files)`
+    );
+    //#endregion
+
+    let compilationFailed: boolean = false;
+    if (diagnostics.length > 0) {
+      this._terminal.writeErrorLine(
+        `Encountered ${diagnostics.length} TypeScript error${diagnostics.length > 1 ? 's' : ''}:`
+      );
+      for (const diagnostic of diagnostics) {
+        this._printDiagnosticMessage(ts, diagnostic, true);
+      }
+
+      compilationFailed = true;
+    }
+
+    if (tslintResult?.failures?.length) {
+      this._terminal.writeWarningLine(
+        `Encountered ${tslintResult!.failures.length} TSLint error${
+          tslintResult!.failures.length > 1 ? 's' : ''
+        }:`
+      );
+      for (const tslintFailure of tslintResult!.failures) {
+        const buildFolderRelativeFilename: string = path.relative(
+          this._configuration.buildFolder,
+          tslintFailure.getFileName()
+        );
+        const { line, character } = tslintFailure.getStartPosition().getLineAndCharacter();
+        const severity: string = tslintFailure.getRuleSeverity().toUpperCase();
+        this._terminal.writeWarningLine(
+          '  ',
+          Colors.yellow(`${severity}: ${buildFolderRelativeFilename}(${line + 1},${character + 1})`),
+          Colors.white(' - '),
+          Colors.yellow(`(${tslintFailure.getRuleName()}) ${tslintFailure.getFailure()}`)
+        );
+      }
+    }
+
+    if (compilationFailed) {
+      throw new Error('TypeScript compilation failed.');
+    }
+  }
+
+  private _printDiagnosticMessage(
+    ts: ExtendedTypescript,
+    diagnostic: Typescript.Diagnostic,
+    withIndent: boolean = false
+  ): void {
+    let terminalMessage: (string | IColorableSequence)[];
+    // Code taken from reference example
+    if (diagnostic.file) {
+      const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!);
+      const message: string = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
+      const buildFolderRelativeFilename: string = path.relative(
+        this._configuration.buildFolder,
+        diagnostic.file.fileName
+      );
+      terminalMessage = [
+        Colors.red(`ERROR: ${buildFolderRelativeFilename}(${line + 1},${character + 1})`),
+        Colors.white(' - '),
+        Colors.gray(`(TS${diagnostic.code})`),
+        ' ',
+        Colors.red(message)
+      ];
+    } else {
+      terminalMessage = [ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n')];
+    }
+
+    if (withIndent) {
+      terminalMessage = ['  ', ...terminalMessage];
+    }
+
+    switch (diagnostic.category) {
+      case ts.DiagnosticCategory.Error: {
+        this._terminal.writeErrorLine(...terminalMessage);
+        break;
+      }
+
+      case ts.DiagnosticCategory.Warning: {
```

Comment:
A major annoyance with our current toolchain is that it does not distinguish "errors" from "warnings" very well (for both ESLint and TypeScript).

Consider this input:
```ts
function example(): void {
  let x = new ForgottenImport();
  let p = new Promise(() => { });
}
```
Heft's output looks like this:
```
[typescript] Encountered 3 TypeScript errors:
[typescript]   ERROR: src\index.ts(5,7) - (TS6133) 'x' is declared but its value is never read.
[typescript]   ERROR: src\index.ts(5,15) - (TS2304) Cannot find name 'ForgottenImport'.
[typescript]   ERROR: src\index.ts(6,7) - (TS6133) 'p' is declared but its value is never read.
[typescript] Encountered 5 TSLint errors:
[typescript]   ERROR: src\index.ts(6,29) - (no-empty) block is empty
[typescript]   ERROR: src\index.ts(5,7) - (prefer-const) Identifier 'x' is never reassigned; use 'const' instead of 'let'.
[typescript]   ERROR: src\index.ts(6,7) - (prefer-const) Identifier 'p' is never reassigned; use 'const' instead of 'let'.
[typescript]   ERROR: src\index.ts(5,7) - (typedef) expected variable-declaration: 'x' to have a typedef
[typescript]   ERROR: src\index.ts(6,7) - (typedef) expected variable-declaration: 'p' to have a typedef
```

This error is a serious mistake, such that the code is unlikely to work at all:
```
[typescript]   ERROR: src\index.ts(5,15) - (TS2304) Cannot find name 'ForgottenImport'.
```

But it is buried in a haystack of other 5 alarm "ERROR" messages complaining about absolutely trivial details, that are totally normal while a person is debugging the code.  Arguably those should really be modeled as "warnings".